### PR TITLE
resource-manager: catch containers earlier when they are gone.

### PR DESCRIPTION
**Pull request by klihub**
**Note 1:** This a stacked PR. You probably want to review the following underlying PRs separately first:
- [x] [#615: topology-aware: better and more readable logs](https://github.com/intel/cri-resource-manager/pull/615)
- [x] [#616: topology-aware: memory accounting and memset expansion fixes](https://github.com/intel/cri-resource-manager/pull/616)

**Note 2:** This PR is also merged into stacked [PR #609](https://github.com/intel/cri-resource-manager/pull/609).

This PR adds code to intercept CRI `StopPodSandbox()` and `ListContainers()`
requests to detect better/earlier when a container is not longer running and to
release the resources of such a container.

### CODE REVIEW

File: pkg/cri/resource-manager/requests.go
Comment (marquiz) [inline pkg/cri/resource-manager/requests.go:236]: Just wondered is it safe to release if the request fails or would it be better to wait for re-try? 🤔  E.g. the container runtime was not running at that very moment and didn't get the request. Kubelet tries again later at which point we've already lost track of the containers.

File: pkg/cri/resource-manager/requests.go
Comment (marquiz) [inline pkg/cri/resource-manager/requests.go:535]: I guess here we could use `fallthrough`(?) 😄

File: pkg/cri/resource-manager/requests.go
Comment (marquiz) [inline pkg/cri/resource-manager/requests.go:536]: I guess here one could use fallthrough
```
		switch {
		case f.Id != "",
		case f.State != nil,
		case f.PodSandboxId != "",
		case len(f.LabelSelector) > 0:
			return reply, nil
		}
```

but I don't know how much difference it really makes

Comment (klihub) [reply pkg/cri/resource-manager/requests.go:536]: Hmm... I was under the impression that a case list is just syntactic sugar and it compiles to a similar/exactly same if-else if-else if construct than a normal switch without case lists.
 
And I guess this is waaay off-topic, but is that really called a `fallthrough` ? I was under the impression that a fallthrough refers to the explicit usage of the `fallthrough` keyword in golang, since unlike in C, by default the cases in a switch construct break instead of falling through.

But your suggestion has the benefit of emphasizing the identical return values in all those cases, so it's probably better in this case.

File: pkg/cri/resource-manager/requests.go
Comment (klihub) [reply pkg/cri/resource-manager/requests.go:236]: Hmm... It's a valid question and I'm not sure which is a/the better way to go.

My approach was simply based on the observation that kubelet will not try to stop a pod unless it thinks all the containers in the pod are already stopped/have exited. Hence, it seemed/seems like a perfect synchronization point to make sure that we are of the same opinion.

Or, to look at it form an alternative perspective, we really don't know why the runtime failed to stop the pod. If it did because the pod does not exist any more, but both the kubelet and us still beg to differ, then ATM we have an effective allocation of resources for something which is not there. That allocation must eventually be (internally) released to prevent it from becoming a resource leak. And if the pod has multiple containers, we potentially have a bunch of those at this point.

I can change this to bail out on error without releasing the containers. Then (assuming that kubelet will not try to indefinitely stop/remove pods if those operations keep failing in the runtime) we have to use intercepted  `ListContainers()` with more complex logic than the current one (which simply releases containers that has recently exited). For some reason I just have a very much Matt Kowalski / Mardi Gras '87 feeling about that...

File: pkg/cri/resource-manager/requests.go
Comment (marquiz) [reply pkg/cri/resource-manager/requests.go:536]: > Hmm... I was under the impression that a case list is just syntactic sugar and it compiles to a similar/exactly same if-else if-else if construct than a normal switch without case lists.

Yes, I think this is the way it is. The idea behind my suggestion was just eliminating identical returns
 
> And I guess this is waaay off-topic, but is that really called a `fallthrough` ? I was under the impression that a fallthrough refers to the explicit usage of the `fallthrough` keyword in golang, since unlike in C, by default the cases in a switch construct break instead of falling through.

Yup, you're correct: fall-through only is the explicit use of `fallthrough` in one case, not combining multiple ifs in one case as a list. My terminology was bad, thanks for correcting me. 

> But your suggestion has the benefit of emphasizing the identical return values in all those cases, so it's probably better in this case.

This was the purpose of my nitty-gritty-suggestion, improving readability 😺  I wasn't sure if was good or bad, though.

Comment (klihub) [reply pkg/cri/resource-manager/requests.go:536]: Okay, so that is syntactically incorrect go. What you can do is this:
```
	switch {
	case f.Id != "",
		f.State != nil,
		f.PodSandboxId != "",
		len(f.LabelSelector) > 0:
		    return reply, nil
	}
```
but that gets turned by the tabulation and formatting nazis into this:
```
	switch {
	case f.Id != "",
		f.State != nil,
		f.PodSandboxId != "",
		len(f.LabelSelector) > 0:
		return reply, nil
	}
```
which is not very readable. So I gave in a replaced it with an `if`.

Comment (marquiz) [reply pkg/cri/resource-manager/requests.go:536]: Woops, yes, my code example was borked 😬  And, nice work gofmt 😢  Maybe an if is actually the best here, agree

File: pkg/cri/resource-manager/requests.go
Comment (marquiz) [reply pkg/cri/resource-manager/requests.go:237]: I fully trust your judgement and reasoning here. Let's keep it the way you wrote it. Sounds safer to me that way. Anyway, my question was just something that occurred to me while reading the code

**Comment by marquiz** [comment]
LGTM 👍  But only after the dependencies have been merged...

File: pkg/cri/resource-manager/requests.go
Comment (klihub) [reply pkg/cri/resource-manager/requests.go:237]: Well, I already went ahead and updated it to not release containers on a failed pod stop request and added nuking to the processing of intercepted list request. Needs to bang that for a while to see if my worst fears are eased enough or if I need to roll back to the original simpler implementation.

File: pkg/cri/resource-manager/requests.go
Comment (kad) [inline pkg/cri/resource-manager/requests.go:216]: should it handle errors like "already stopped" or similar conditions?

Comment (klihub) [reply pkg/cri/resource-manager/requests.go:216]: If you mean if handling errors returned by the runtime like 'already stopped' for StopPodSandbox or StopContainer with a stopped pod's of container's ID, then there are no errors like that. The runtime will reply with success if you try to stop an already stopped container or pod.

**Review by askervin — APPROVED** [review]
LGTM

**Review by marquiz — APPROVED** [review]
lgtm

**Pull request merged**
